File: src/main/java/org/taktik/icure/logic/impl/ClassificationLogicImpl.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic.impl;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.taktik.icure.dao.ClassificationDAO;
import org.taktik.icure.dao.impl.idgenerators.UUIDGenerator;
import org.taktik.icure.entities.Classification;
import org.taktik.icure.entities.embed.Delegation;
import org.taktik.icure.logic.ClassificationLogic;
import org.taktik.icure.logic.ICureSessionLogic;
import org.taktik.icure.validation.aspect.Check;

import javax.validation.constraints.NotNull;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;


/**
 * Created by dlm on 16-07-18
 */
@org.springframework.stereotype.Service
public class ClassificationLogicImpl extends GenericLogicImpl<Classification, ClassificationDAO> implements ClassificationLogic {
	private static final Logger log = LoggerFactory.getLogger(ClassificationLogicImpl.class);


	private ClassificationDAO classificationDAO;
	private UUIDGenerator uuidGenerator;
	private ICureSessionLogic sessionLogic;

	@Autowired
	public void setClassificationDAO(ClassificationDAO classificationDAO) {
		this.classificationDAO = classificationDAO;
	}
	@Autowired
	public void setUuidGenerator(UUIDGenerator uuidGenerator) {
		this.uuidGenerator = uuidGenerator;
	}
	@Autowired
	public void setSessionLogic(ICureSessionLogic sessionLogic) {
		this.sessionLogic = sessionLogic;
	}
	@Override
	protected ClassificationDAO getGenericDAO() {
		return classificationDAO;
	}

	@Override
	public Classification createClassification(@Check @NotNull Classification classification) {
		List<Classification> createdClassifications = new ArrayList<>(1);
		try {
			// Fetching the hcParty
			String healthcarePartyId = sessionLogic.getCurrentHealthcarePartyId();

			// Setting Classification attributes
			classification.setId(uuidGenerator.newGUID().toString());
			classification.setAuthor(healthcarePartyId);
			classification.setResponsible(healthcarePartyId);

			createEntities(Collections.singleton(classification), createdClassifications);
		} catch (Exception e) {
			log.error("createClassification: " + e.getMessage());
			throw new IllegalArgumentException("Invalid Classification", e);
		}
		return createdClassifications.size() == 0 ? null:createdClassifications.get(0);
	}

	@Override
	public Classification getClassification(String classificationId) {
		return classificationDAO.getClassification(classificationId);
	}

	@Override
	public List<Classification> findByHCPartySecretPatientKeys(String hcPartyId, List<String> secretPatientKeys) {
		return classificationDAO.findByHCPartySecretPatientKeys(hcPartyId, secretPatientKeys);
	}

	@Override
	public Set<String> deleteClassifications(Set<String> ids) {
		try {
			deleteEntities(ids);
		} catch (Exception e) {
			log.error(e.getMessage(), e);
			return null;
		}
		return ids;
	}

	@Override
	public Classification modifyClassification(@Check @NotNull Classification classification) {
		try {
			Classification toEdit = getClassification(classification.getId());
			toEdit.setLabel(classification.getLabel());
			updateEntities(Collections.singleton(toEdit));
            return getClassification(classification.getId());
		} catch (Exception e) {
			throw new IllegalArgumentException("Invalid Classification", e);
		}
	}

	@Override
	public Classification addDelegation(String classificationId, String healthcarePartyId, Delegation delegation) {
		Classification classification = getClassification(classificationId);
		classification.addDelegation(healthcarePartyId, delegation);

		return classificationDAO.save(classification);
	}

	@Override
	public Classification addDelegations(String classificationId, List<Delegation> delegations) {
		Classification classification = getClassification(classificationId);
		delegations.forEach(d->classification.addDelegation(d.getDelegatedTo(),d));

		return classificationDAO.save(classification);

	}

    @Override
    public List<Classification> getClassificationByIds(List<String> ids) {
        return classificationDAO.getList(ids);
    }

}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/ClassificationLogicImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.asynclogic.impl

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOf
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.taktik.couchdb.DocIdentifier
import org.taktik.couchdb.id.UUIDGenerator
import org.taktik.icure.asyncdao.ClassificationDAO
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.ClassificationLogic
import org.taktik.icure.entities.Classification
import org.taktik.icure.entities.embed.Delegation

/**
 * Created by dlm on 16-07-18
 */
@Service
class ClassificationLogicImpl(
	private val classificationDAO: ClassificationDAO,
	private val uuidGenerator: UUIDGenerator,
	private val sessionLogic: AsyncSessionLogic
) : GenericLogicImpl<Classification, ClassificationDAO>(sessionLogic), ClassificationLogic {

	override fun getGenericDAO(): ClassificationDAO {
		return classificationDAO
	}

	override suspend fun createClassification(classification: Classification) = fix(classification) { classification ->
		try { // Fetching the hcParty
			val userId = sessionLogic.getCurrentUserId()
			val healthcarePartyId = sessionLogic.getCurrentHealthcarePartyId()
			createEntities(
				setOf(
					classification.copy(
						author = userId,
						responsible = healthcarePartyId
					)
				)
			).firstOrNull()
		} catch (e: Exception) {
			log.error("createClassification: " + e.message)
			throw IllegalArgumentException("Invalid Classification", e)
		}
	}

	override suspend fun getClassification(classificationId: String): Classification? {
		return classificationDAO.getClassification(classificationId)
	}

	override fun listClassificationsByHCPartyAndSecretPatientKeys(hcPartyId: String, secretPatientKeys: List<String>): Flow<Classification> = flow {
		emitAll(classificationDAO.listClassificationsByHCPartyAndSecretPatientKeys(hcPartyId, secretPatientKeys))
	}

	override fun deleteClassifications(ids: Set<String>): Flow<DocIdentifier> {
		return try {
			deleteEntities(ids)
		} catch (e: Exception) {
			log.error(e.message, e)
			flowOf()
		}
	}

	override suspend fun modifyClassification(classification: Classification) = fix(classification) { classification ->
		try {
			classification.id.let {
				getClassification(it)?.let { toEdit ->
					modifyEntities(setOf(toEdit.copy(label = classification.label))).firstOrNull()
				}
			} ?: throw IllegalArgumentException("Non-existing Classification")
		} catch (e: Exception) {
			throw IllegalArgumentException("Invalid Classification", e)
		}
	}

	override suspend fun addDelegation(classificationId: String, healthcarePartyId: String, delegation: Delegation): Classification? {
		val classification = getClassification(classificationId)
		return classification?.let {
			classificationDAO.save(
				it.copy(
					delegations = it.delegations + mapOf(
						healthcarePartyId to setOf(delegation)
					)
				)
			)
		}
	}

	override suspend fun addDelegations(classificationId: String, delegations: List<Delegation>): Classification? {
		val classification = getClassification(classificationId)
		return classification?.let {
			return classificationDAO.save(
				it.copy(
					delegations = it.delegations +
						delegations.mapNotNull { d -> d.delegatedTo?.let { delegateTo -> delegateTo to setOf(d) } }
				)
			)
		}
	}

	override fun getClassifications(ids: List<String>): Flow<Classification> = flow {
		emitAll(classificationDAO.getEntities(ids))
	}

	companion object {
		private val log = LoggerFactory.getLogger(ClassificationLogicImpl::class.java)
	}
}
